### PR TITLE
[model checkpoint] CPU communication ops

### PR DESCRIPTION
**Pull request by kurisusnowdeng**
- Updated communication ops for cpu tensors
- Added scatter_object_list from torch

### CODE REVIEW

File: colossalai/communication/collective.py [hunk 1]
```diff
@@ -8,10 +8,13 @@
...
 from colossalai.context import ParallelMode
 from colossalai.core import global_context as gpc
-from colossalai.utils import get_current_device
 
 
-def all_gather(tensor: Tensor, dim: int, parallel_mode: ParallelMode, async_op: bool = False) -> Tensor:
+def all_gather(tensor: Tensor,
+               dim: int,
+               parallel_mode: ParallelMode,
+               on_cpu: bool = False,
```
Comment (feifeibear) [inline colossalai/communication/collective.py:16]: on_cpu may have a conflict with tensor.device.
we have to ensure tensor.device is CPU?

File: colossalai/communication/collective.py [hunk 2]
```diff
@@ -54,6 +59,7 @@ def reduce_scatter(tensor: Tensor,
                    dim: int,
                    parallel_mode: ParallelMode,
                    op: ReduceOp = ReduceOp.SUM,
+                   on_cpu: bool = False,
```
Comment (feifeibear) [inline colossalai/communication/collective.py:62]: on_cpu may have a conflict with tensor.device.
we have to ensure tensor.device is CPU?

**Comment by feifeibear** [comment]
#256

File: colossalai/communication/collective.py (hunk 1 above)
Comment (kurisusnowdeng) [reply colossalai/communication/collective.py:16]: Let me fix. I think maybe we do not need this option, but just check if the tensor is on cpu.

**Review by feifeibear — APPROVED** [review]

**Pull request merged**